Added option for C++ standard

kunjsong01 opened the pull request:
~~Let's work with AST generated by modern C++14 standard.~~ 
~~Force standard to be C++14 standard based on comments in https://github.com/esbmc/esbmc/issues/972~~

Edit: added an option for C++ standard. Default is C++03.

kunjsong01 committed:
[clang-cpp-convert] migrate to c++14 standard

kunjsong01 committed:
[clang-cpp-frontend] skipped move ctor

kunjsong01 committed:
[regression] disabled TCs using move semantics

rafaelsamenezes commented on src/clang-cpp-frontend/clang_cpp_language.cpp:
```diff
@@ -21,8 +21,8 @@ languaget *new_clang_cpp_language()
-  // We also force the standard to be c++98
-  compiler_args.push_back("-std=c++98");
+  // We also force the standard to be c++14
+  compiler_args.push_back("-std=c++14");
```
This could be the default, but you should add a flag to set C++ standard as flag. Then you could duplicate the testcases with a C++98 passing and a C++14 KNOWNBUG

lucasccordeiro replied:
I agree with @rafaelsamenezes' suggestion.

kunjsong01 replied:
Done.

lucasccordeiro reviewed (approved):
Please consider @rafaelsamenezes' suggestion.

kunjsong01 committed:
[clang-cpp-frontend] reverted C++14 standard

  - We should add a flag for C++ language standard, instead of
    forcing it to a specific one.

kunjsong01 committed:
[clang-cpp-frontend] added a flag for C++ std

rafaelsamenezes commented on src/esbmc/options.cpp:
```diff
@@ -72,6 +72,9 @@ const struct group_opt_templ all_cmd_options[] = {
     {"warning,W",
      boost::program_options::value<std::vector<std::string>>(),
      ""},
+    {"cppstd",
```
In general, we consider flags with only one "-" to be passed through the frontend. Also, users integrate ESBMC by using a build system so they would prefer to pass `-std=c++11` than `--cppstd=11`. I would suggest to also look at the flags that we pass to the frontends.

kunjsong01 replied:
logged in https://github.com/esbmc/esbmc/issues/1050, which shall be done in another PR.

rafaelsamenezes commented on src/esbmc/options.cpp:
```diff
@@ -72,6 +72,9 @@ const struct group_opt_templ all_cmd_options[] = {
      ""},
+    {"cppstd",
+     boost::program_options::value<std::string>()->value_name("version"),
+     "set C++ standard"},
```
I would change the message to `set C++ standard (available: 98, 03, 11, 14, or 17)`

kunjsong01 replied:
good point.
Added

lucasccordeiro reviewed (approved):
LGTM.

Many thanks, @kunjsong01.

kunjsong01 committed:
[regression] added TCs

kunjsong01 committed:
[options] added standatd

fbrausse reviewed (commented):
I agree with @rafaelsamenezes' comment regarding `-std=c++11`. Also, to check whether this switch actually makes a difference, I would suggest adding a test case like `regression/esbmc-cpp/bug_fixes/972_std14/main.cpp` but with `--cppstd 98` expecting a parsing error.

kunjsong01 committed:
[regression] fixed TCs and added a TC using c++98

kunjsong01 commented:
> I would suggest adding a test case like regression/esbmc-cpp/bug_fixes/972_std14/main.cpp but with --cppstd 98 expecting a parsing error.

Good point. I've added another test case using c++98

kunjsong01 commented:
Regarding the option styling, `--cppstd 98` is pretty much in line with other frontend options, and other options taking a numerical value e.g. "--k-step nr" in ESBMC. There are 39 frontend options in total. 35 out of 39 options have the two hyphens "--" style.

[list_of_frontend_options.txt](https://github.com/esbmc/esbmc/files/11614793/list_of_frontend_options.txt)

The `-std=c++11` style is in line with the compilers. But we are not writing a compiler. As for the users, I don't see `--cppstd 11` is less intuitive than `-std=c++11` in any way.  

Do we have any existing option in the format `-option=<number>` ? If not, for now an issue might be just enough for aesthetic improvement?

fbrausse commented:
> The `-std=c++11` style is in line with the compilers. But we are not writing a compiler.

I tend to disagree on this stance: ESBMC has options that are known from (non-Windows) compilers, such as `-I`, `-D`, `--sysroot` even though some compilers only allow a single dash for the long options. I regard `-std=c++N` or `--std c++N` similarly universal. Whether it is one dash with an `=` symbol or two dashes and space separated in my opinion is no big deal.

> As for the users, I don't see `--cppstd 11` is less intuitive than `-std=c++11` in any way.

The difference between these two is the semantics the option has: do we want a similar option for each frontend language that supports multiple versions (C, in particular), or should the language and standard be part of the same single `-std` option? Some time ago @mikhailramalho said that atm. ESBMC does "not really" support multiple different input languages and if that fact remains, I would really opt for a `-std` style option. Even if not, I'd rather vote for context- (aka position-) sensitive option processing akin `-std=c99 file1.c -std=c++98 file2.cc` instead of specifying a standard language-wide.

kunjsong01 committed:
[regression] fixed 972 TC

fbrausse reviewed (approved):
LGTM.

kunjsong01 commented:
>The difference between these two is the semantics the option has: do we want a similar option for each frontend language that supports multiple versions (C, in particular), or should the language and standard be part of the same single -std option?

I see... makes sense, indeed. 

If we opt for `-std` style, it would require a complete re-writing of this patch.

BTW, it doesn't seem straightforward with [boost program optoin without writing a bespoke parser](https://stackoverflow.com/questions/19422541/boost-program-option-force), unless we write a bespoke parser. 

Do we have any existing option in the format `-option=<number>`?  If so, it should be just a implementation with a snap. 
Otherwise, https://github.com/esbmc/esbmc/issues/1050 should do, and it's probably better to push this patch, as we have an external user relying on it.

fbrausse commented:
> If we opt for `-std` style, it would require a complete re-writing of this patch.

Indeed. Fortunately it's not invasive.

> Do we have any existing option in the format `-option=<number>`?

I'm not aware of any. Just go forward with this PR until #1050 sorts out how to generalize this to other languages.
